feat: Network log/Journal reading module tradegame.py

kfsone opened the pull request:
The experiment was: What if TD had some minimal support for the network journal?

This isn't intended to be an all-bells-and-whistles component; it reads the journals somewhat naively, doesn't provide a billion methods for accessing this and that.

I stop at some trade-useful data: what's your ship and where is it at; if your thing needs more, the data is here in dict format.

For now it consists of some hacks in the "trade" sub command:
`~` for current system/station, and `~@` for nav-target system,

```
trade.py trade ~ "~@/stronghold carrier"
```

If you have a nav target, this would try and find trades between your currently-docked station and the stronghold carrier your nav target system.

```
trade.py import -P journal
```
will read in the prices from your last visited commodities market.

kfsone committed:
Incorporate basic game log file reading

feat: add tradegame.py and EliteGame class for reading journal/network log
- Designed to be extensible with only minimal journal knowledge,
- Uses `orjson` for super-fast parsing/load times,
- Crude understanding of cargo capacity,
- Crude understanding of current system/station,
- Exposes journal/log file data as dictionaries without curation,
- Default paths for Windows and MacOS; use ELITE_JOURNAL_PATH environment var to set/override

feat: 'journal' import plugin

`trade.py import -P journal` will import the last market prices you saw in-game if they are
newer than the ones in the database.

refactor: add non-tradedb methods for duration normalization

feat: allow trade_cmd to read data from the game
preliminary syntax:
- When docked, "~" = current system/current station,
- When not docked, use "~/station name",
- With a route selected, "~@" expands to target *system* name

feat: add --fill, --load, --full-load to trade command

these (mutually-exclusive) options all require live game data access to read your cargo capacity and load.

"--fill":
for each available item, show the maximum profit you could make based on supply and
maximum cargo capacity:
```
Win|pwsh> python trade.py trade ~ ~@/hole --fill
7 trades found between Kamitra/Hammel Terminal and Kuwemaki/The Jet's Hole.
 Item              Profit       Cost     Buying |  Units      Profit
--------------------------------------------------------------------
 Water                686         22        708 |    108      74,088
 Fish                 566        293        859 |     30      16,980
 Wine                 527        162        689 |     55      28,985
 Grain                517        122        639 |     74      38,258
 Beer                 513         92        605 |     81      41,553
 Liquor               420        474        894 |     12       5,040
 Hydrogen Fuel         52         88        140 |    108       5,616
````

"--load":
- fetches cargo capacity from game journal,
- fetches current cargo quantity from game journal,
- lists the most profitable items with sufficient supply/demand until cargo hold is full,

"--full-load":
- as "--load" but ignores your current cargo (e.g if you're going to sell everything)

Examples, assuming you have 50 units of cargo in a hold with 251 capacity:

```
trade.py trade "Loha/Bell" "Byzan/Const"
... shows all items tradeable Bell Plant -> Constantinople

trade.py trade "Loha/Bell" "Byzan/Const" --load -vv
7 trades found between Loha/Bell Plant and Byzantine/Constantinople.
 Item                                            Profit       Cost     Buying     Supply     Demand    SrcAge    DstAge |  Units      Profit
--------------------------------------------------------------------------------------------------------------------------------------------
 Chemicals/Gold                                  51,581      4,945     56,526          1    103,828      1.7M     67.9M |      1      51,581
 Chemicals/Silver                                45,337      4,089     49,426          8    415,091      1.7M     67.9M |      8     362,696
 Chemicals/Beryllium                             39,133      3,959     43,092          2     48,168      1.7M     67.9M |      2      78,266
 Chemicals/Gallium                               24,888      2,479     27,367        384    247,010      1.7M     67.9M |    190   4,728,720
                                                                                     ^^^-------------------------------------^^^
```

here there are more items available, (supply of Gallium = 382) but it only counted enough *units* to fill your hold.

However, if you are going to sell the 50 units already in your hold first, --full-load would produce:

```
trade.py trade "Loha/Bell" "Byzan/Const" --full-load -vv
7 trades found between Loha/Bell Plant and Byzantine/Constantinople.
 Item                                            Profit       Cost     Buying     Supply     Demand    SrcAge    DstAge |  Units      Profit
--------------------------------------------------------------------------------------------------------------------------------------------
 Chemicals/Gold                                  51,581      4,945     56,526          1    103,828      1.7M     67.9M |      1      51,581
 Chemicals/Silver                                45,337      4,089     49,426          8    415,091      1.7M     67.9M |      8     362,696
 Chemicals/Beryllium                             39,133      3,959     43,092          2     48,168      1.7M     67.9M |      2      78,266
 Chemicals/Gallium                               24,888      2,479     27,367        384    247,010      1.7M     67.9M |    240   5,973,120
                                                                                     ^^^-------------------------------------^^^
```

"--load" and "--full-load" include a total unless you use "-q":
```
Win|pwsh> python trade.py trade "Zlotrimi/Alvarez de Pineda Prospect" "LB 3303/Poindexter Horizons" --full-load -v
2 trades found between Zlotrimi/Alvarez de Pineda Prospect and LB 3303/Poindexter Horizons.
 Item                Profit       Cost     Buying    SrcAge    DstAge |  Units      Profit
------------------------------------------------------------------------------------------
 Metals/Gold         25,076     41,914     66,990        7D       10D |    773  19,383,748
 Metals/Tantalum      2,297     12,527     14,824        7D        8D |    315     723,555
------------------------------------------------------------------------------------------
Total Units: 1,088. Total Profit: 20,107,303.
```

style: various hinting/cleanup

kfsone commented:
I'm feeling like this is mostly ready to go, even if we leave the `~` and `~@` nouns as experimental, maybe mark the `offload` command as _experimental_. Or just remove it entirely.

Although I've got to admit I've been using the heck out of everything in this branch

```
10  python trade.py offload  # help me dump what I have
# land & sell,
20  python trade.py import -P journal ; python trade.py import -P eddblink -O 7days,units,skipvend,optimize
# caught up
30  python trade.py trade ~ ~@/whereimgoing
# sell, take missions, run missions, acquire random bits of cargo, 
40  goto 10
```

kfsone commented:
copilot review for the giggles: Unfortunately, TD is a significant source of language modelling for copilot, so it gives _really_ bad feedback :(

kfsone commented on tradedangerous/commands/offload_cmd.py:
```diff
@@ -0,0 +1,723 @@
+
+        for stat in station_stats:
+            station = stat['station']
+            max_age_days = max(max(now - row.modified for row in stat['items']).seconds, 0)
```
.days returns an integer.

kfsone commented:
I didn't go to town adding use of this to commands, because that really wants doing in the TradeORM lookup code, so that the support is there and global rather than per-tool, or it should be done in the commandenv code. So having the journal support int he first place is step 1.

Tromador commented:
I will give you a proper review later, but this stands out 
```
def native_elite_folder() -> Path:
    """ native_elite_folder discovers the default Elite: Dangerous saved games folder for the current OS or raises a RuntimeException. """
    match sys.platform:
        case "win32" | "cygwin":
            return WINDOWS_JOURNAL_DIR
        case "darwin":
            return MACOS_JOURNAL_DIR
        case _:
            raise GameException(f"Unsupported operating system: {sys.platform}. Please specify elite_game_path parameter explicitly.")
```

What about people playing on Linux via Proton or some other wine implementation? There's actually a lot of people who do and it's only going to increase as steam's Linux support is really good and stuff like the steam cube is released. Can we not get a default path for that? 

Assuming we can't and hence your error message says "Please give us the path", it's still a problem because you run this function before checking for an explicit path, so even if our Linux user gives a path, EliteGame.__init_ runs this function before looking at the provided path, so it still bombs out. It's just in the wrong order.

More to follow after I sleep.

Tromador commented:
### `EliteGame.list_journals() optional regex group → TypeError`

`EliteGame.list_journals()`scans files named like `Journal.YYYY-MM-DDTHHMMSS(.NN).log` and builds a sortable “index” string from the captured groups so it can sort journals newest-first. So far so good. The nit is that in the regex the last capture `(\.\d+)` is optional because it’s wrapped in `(?:(\.\d+))?`.

If the .NN is actually optional, and there's a file where it's missing, `match_groups()` will return a tuple with a `None` type which will then blow up as a `TypeError` when you try and `join()` it later. Conversely if the .NN is always there, then the regex shouldn't say it's optional. I can't prove either way, but all my journals have .NN in their filenames, so unless you know otherwise, I suggest it's always going to be there.

### `journal_plug` exception handling doesn’t match `EliteGame` behaviour

In `journal_plug` you're looking for an exception on file not found.

```
try:
    game = EliteGame(json_files=[JsonFiles.MARKET])
except FileNotFoundError:
    self.tdenv.NOTE("No Market.json present, cancelling.")
    return False

market: dict[str, typing.Any] = game.json_data[JsonFiles.MARKET]
```

But `elite_game` doesn't raise that for a missing json. It just logs a warning and carries on.

```
if not json_file_path.exists():
    self.tdenv.WARN("json file not found: {}", json_file_path)
    continue
...
self.json_data[json_file_enum] = data
```

Thus “Market.json missing” won’t be a clean cancel; it’ll be a warning from `EliteGame`, followed by a crash in the plugin when it indexes `json_data[...]`.

### `EliteGame(json_files=[MARKET])` still loads journals by default — so journal-related failures can abort a Market-only import

Even though the plugin’s mission statement is “read the live market json”, the call:

tradedangerous/plugins/journal_plug.py:ImportPlugin.run

```
game = EliteGame(json_files=[JsonFiles.MARKET])
```

will still run EliteGame.refresh(), which calls _load_journals() first:

tradedangerous/tradegame.py:EliteGame.refresh

```
with bench(..., "_load_journals"):
    self._load_journals()
with bench(..., "_load_json_files"):
    self._load_json_files()
```

…and _load_journals() will throw if there are no journals:

tradedangerous/tradegame.py:EliteGame._load_journals

```
files = self.list_journals()
if not files:
    raise JournalException(f"no journal logs found in {self.path}")
```

So the plugin can fail for reasons completely unrelated to Market.json, e.g.:

- journal folder exists but contains no journal logs
- journals exist but are mid-write / unreadable (there’s a “please wait…” exception path)
- user wants “just import Market.json” but gets blocked by journal parsing

More to follow...

kfsone commented:
@Tromador refactored the path lookup code, ELITE_JOURNAL_PATH environment var overrides the defaults, and it will try default paths until one matches. I don't know what the defaults are on things like steamdeck or etc so I didn't try adding any.

kfsone commented:
Ok - I just realized this isn't rebased off of release/v1 so there was a whole bunch of unrelated stuff in it -- cleaning up now

Tromador commented:
> @Tromador refactored the path lookup code, ELITE_JOURNAL_PATH environment var overrides the defaults, and it will try default paths until one matches. I don't know what the defaults are on things like steamdeck or etc so I didn't try adding any.

The problem (as it was, maybe you fixed this) is that it tested for defaults and raised if it couldn't find one. Even if the player provided an explicit path. Sure we might not be able to know the default for a particular OS, but it wasn't even allowing the player to tell us where it was. It needs to test for the explicit path and raise if that doesn't exist, then only if the explicit path isn't provided should it attempt to find a default and raise if it can't.

I hope that makes sense, my medications have just smacked me in the face and I'm very wobbly atm.

Tromador commented:
> Ok - I just realized this isn't rebased off of release/v1 so there was a whole bunch of unrelated stuff in it -- cleaning up now

Probably because I merged your other PR last night :)

kfsone commented:
> > @Tromador refactored the path lookup code, ELITE_JOURNAL_PATH environment var overrides the defaults, and it will try default paths until one matches. I don't know what the defaults are on things like steamdeck or etc so I didn't try adding any.
> 
> The problem (as it was, maybe you fixed this) is that it tested for defaults and raised if it couldn't find one. Even if the player provided an explicit path. Sure we might not be able to know the default for a particular OS, but it wasn't even allowing the player to tell us where it was. It needs to test for the explicit path and raise if that doesn't exist, then only if the explicit path isn't provided should it attempt to find a default and raise if it can't.
> 
> I hope that makes sense, my medications have just smacked me in the face and I'm very wobbly atm.

Absolutely - I retooled that bit entirely - put it in a helper that starts with the most effective first and works down from that:

- the explicit 'path' argument passed to the parent method,
- tdenv "journal_path" property (the names were annoyingly long and didn't make sense outside the module),
- environment variable (named constant so it can tell the user which one),
- windows or macos default path (and it tries both, incase of emulation, etc)
- else throw a guidance-providing exception

I'm going to move the pyproject stuff into a separate branch to minimize concerns, since it turns out it's not uv specific, it's just how modern python does it, and it happens to benefit anyone using uv also.

Tromador commented:
So in addition to the stuff I already flagged:

### `trade --fill` can `TypeError` if cargo capacity wasn’t derived

trade_cmd.py has 
```
results.summary.cargo_space = 1
if game and getattr(cmdenv, "fill", False):
    results.summary.cargo_space = max(game.get_status().cargo_space, 1)
```
 but tradegame.py has
 ```
@dataclass
class Status:
    ...
    cargo_space: int | None = None
```
In practice this might never be a problem. Everyone should have `ModulesInfo.json` and it should sort itself, but the code allows `None` and it might be worth guarding for it.

### `~@/...` can `KeyError` if `NavRoute.json` isn’t present

Again, I couldn't say if the file will ALWAYS be there by definition, but you have in `trade_cmd.py:lookup_game_name`
```
nav_data = game.json_data[JsonFiles.NAVROUTE] or {}
```
and then in `tradegame.py:EliteGame._load_json_files`
```
if not json_file_path.exists():
    self.tdenv.WARN("json file not found: {}", json_file_path)
    continue
```
So if `NavRoute.json` doesn’t exist, `NAVROUTE` won’t be a key in `json_data`, and the indexing form will `KeyError` before your `or {}` ever runs.

So again, you might want to guard for this, if it's even slightly possible this file might not be present.

There's a few other nits and stuff, but I think I've covered all the big stuff. If you can have a look at those, then I'll go over it again.

kfsone commented:
- addressed .LOGN issue,
- addressed exceptions:
- addressed upto-dateness being an "error"
<img width="2980" height="323" alt="image" src="https://github.com/user-attachments/assets/d057db2f-bc26-46b7-b4bb-7d3d0feea2f3" />
---
<img width="2987" height="305" alt="image" src="https://github.com/user-attachments/assets/7b5b12e9-3f0e-4dd7-afcb-69989d2e08d6" />
---
<img width="2999" height="299" alt="image" src="https://github.com/user-attachments/assets/cf8a2a13-ced4-40b9-b9aa-5eccbcde2862" />
---
<img width="1568" height="166" alt="image" src="https://github.com/user-attachments/assets/c63f92c7-847d-4893-8ce3-32662dcb84e3" />
* spillover: we'll need "describeAge" to come out of tradedb.py soon, so I moved it to misc.tradeage.describe_age

Yes - I use snake case now instead of camel when writing Python ;-P

Still working on above feedback.

kfsone commented:
- addressed cargo_space = None
- ~@ keyerror - addressed above (exception with guidance)

I've left this as commits so you can see the substance, I'll squash when we're good.

Tromador commented:
You've given yourself a regression dude.

`journal_plug`calls `self.tdenv.note(...)` (lowercase) which will blow up

`TradeEnv.__getattr__` only synthesises uppercase log methods like `NOTE`, not lowercase `note`, and returns `None` for unknown attributes. So `self.tdenv.note(...)` becomes “call None”.

In `plugins/journal_plug.py:ImportPlugin.run`:
```
self.tdenv.note("[green]Updating:[/green] previous entry: [yellow]{}[/]/{}", last_mod, age)
self.tdenv.note("[green]New Data[/green] [dim](station had no recent data)[/dim]")
```
In `tradeenv.py:TradeEnv.__getattr__`:
```
case "NOTE":
theme_prefix, theme_label = self.theme.note, self.theme.NOTE
…
case _:
pass
return None
```
**Impact**: runtime TypeError as soon as those lines execute (i.e. whenever the plugin reaches that branch).

kfsone commented:
damn, good catch. fixed

Tromador commented:
### Stuff that looks fixed
- LOGN / `.NN` requirement. `EliteGame.list_journals` now requires the `.NN` suffix
- “Up-to-dateness” is no longer treated as an error. `journal_plug` now treats it as a note + returns.
- `cargo_space = None` no longer explodes. Trade command clamps it safely now.
- Exceptions now come from `EliteGame` instead of ad-hoc catching in the plugin. 

## Remaining issues.
### `JournalLoad.ALL_JOURNALS` is currently broken by the match order
`ALL_JOURNALS = -1`, but your first match arm treats any `int <= 0` as “don’t load”, so `-1` will never reach the ALL_JOURNALS case.
In tradegame.py:EliteGame._load_journals
```
match self.journal_load:
    case JournalLoad.DO_NOT_LOAD | int() as n if n <= 0:
        self.journals.clear()
        return

    case JournalLoad.ALL_JOURNALS:
        max_journals = -1
```
**Impact**: any call-site that sets `journal_load = JournalLoad.ALL_JOURNALS` (or `-1`) will silently load zero journals.

**Fix**: handle `ALL_JOURNALS` (or `-1`) before the `int() if n <= 0` arm, or narrow that arm to only `0`

### `MissingJsonError` is being used for missing journal files too

In _load_journals, a missing journal log file raises MissingJsonError. You might be using that as a generic "file not found" check, but it's going to raise a misleading error message.
```
except FileNotFoundError as e:
    raise MissingJsonError(str(e)) from None
```

### Trade command now always requests `NavRoute.json` when `--fill` or any `~` is used.

That might be intentional, but the consequence is important: with the new “missing JSON is an exception” behaviour, this makes `NavRoute.json` effectively required in those paths, even when the user is only using `~` (not `~@`).
in trade_cmd.py:run
```
if want_game or want_fill:
    game = EliteGame(tdenv=cmdenv, extra_jsons=[JsonFiles.NAVROUTE])
```
And missing JSON is now a hard exception in `_load_json_files`
**Impact**: user can get a hard failure just by using `--fill` or `~` if `NavRoute` isn’t present (depending on how Elite populates that file). If the intent was “only require NavRoute.json for ~@”, then it could do with tightening.

### `lookup_game_name` uses a walrus inside an `assert` (I had to look up what a walrus is, that's kinda funny).

If Python is run with `-O`, asserts are stripped, meaning `game` is never assigned and you’ll get a `NameError` immediately afterwards. Personally, I've never run it with -O, but it only needs a user to have `PYTHONOPTIMIZE=1` in their environment and this will hit them.

Just make it explicit:
- `game = getattr(cmdenv, "game", None)`
- `if not game: raise CommandLineError(...)`

That preserves the nice error message and works under all interpreter modes.

kfsone commented:
good catches, will address tomorrow!

kfsone commented:
I think I've addressed all the concerns here, the remaining ones are mutual:

TradeORM lookups aren't yet on-par with the all-singing/dancing fuzzy lookups of  TradeDB, so we can either close/gate this PR predicated on a (separate) tradeorm lookup implementation or consider whether we can accept a short-term simplification of how the trade-subcommand handles system/station lookups. I'm finding the journal/data-access features useful enough that this is the branch I routinely operate out of.

Tromador commented:
What's going on with the docs? Doesn't look like you actually changed anything - something weird going on with your local repo?

It even looks to have referenced stuff in "archive" which is "stuff kept for historical reference only".

kfsone commented:
Those files have inconsistent line endings (a mix of crlf and lf). I have a PR that normalizes the line endings to one style: https://github.com/eyeonus/Trade-Dangerous/pull/272

I've already merged that into this PR, that's why they're showing up here.

kfsone commented:
K - this should be clean now.

Tromador commented:
This is getting closer, but it has remaining issues. 

### `JournalLoad.ALL_JOURNALS` is **still** broken by the match order
```
match self.journal_load:
    case JournalLoad.DO_NOT_LOAD | int() as n if n <= 0:
        ...
    case JournalLoad.ALL_JOURNALS:
        max_journals = -1
```
You enum` ALL_JOURNALS` to be `-1` and that is `<=0` so it will hit `DO_NOT_LOAD` and stop there.

### Missing json is raised for missing journals.
Journals are not .json files they are .log files. 
```
with bench(...):
    try:
        self._load_journal_files(files)
    except FileNotFoundError as e:
        raise MissingJsonError(str(e)) from None
```
Which end up with the user getting a misleading error message.

### A subtle but nasty correctness bug introduced in `trade_cmd --load`
You fixed the original `max(None, 1)` TypeError by doing `cargo_space = status.cargo_space or 1`, but `--load` then subtracts cargo load:
```
cargo_space = status.cargo_space or 1
if want_load and not full_load:
    cargo_space -= status.cargo_load  # can make cargo_space negative
results.summary.cargo_space = max(cargo_space, 1)
...
if want_load:
    spare_units = cargo_space - units_seen
    units = min(units, spare_units)   # can drive units negative
```
If `status.cargo_space` is unknown (`None`) but cargo_load is known (`non-None`), `cargo_space` becomes `1 - cargo_load` (negative), but only the summary is clamped — the local `cargo_space` remains negative and can produce **negative “units”** rows.

### `journal_plug` will always load journals, and ignores the command’s env
1: Unless you specifically tell it not to, it **will** load journals
I presume this is by intention, but please just confirm that. 

2: It ignores the command/plugin environment (because it doesn’t pass self.tdenv)
In EliteGame.__init__:
```
self.tdenv = tdenv or TradeEnv()
self.path = journal_path(self.tdenv, path)
```
Since `journal_plug` doesn’t pass `tdenv=self.tdenv`, EliteGame constructs a fresh default `TradeEnv()`. That means:
any CLI/config values living on the plugin’s self.tdenv (e.g. a configured journal path, journal_load override, debug level, etc.) are not used by EliteGame.
even logging/debug output from EliteGame goes through that new default env, not the command’s env.
End result, defaults are hard coded, even if user overrides on command line.

kfsone commented:
Excellent feedback as always.

## all journals
Ah, good catch; I'd used a sentinel object originally and hadn't spotted that,

## missing json raised
Bizzare, fixed

## "do we have it" checks

data checks/consistency: Having looked at migrating a few other commands, your points about field checking are very on-point; so I added a "require_game_data" free-standing helper in tradegame that lets you tell it what you're going to be needing, including events, navroute, and status fields.

They then pass thru a single normalized guidance construct to tell the player "you tried to use game data I don't have, maybe wait til the game is done loading".

_Future think: store data to a db so we don't have to keep trying to look it up/parse it. That way when the game creates a new journal, we're not u.s.c.w.a.p ;)_

## journal plug

Just an oversight, the api grew under the plugin. Fixed.

kfsone commented:
<img width="1077" height="359" alt="image" src="https://github.com/user-attachments/assets/1443c9be-ddbf-4b87-8dbc-e79f9592da1d" />

<img width="1186" height="232" alt="image" src="https://github.com/user-attachments/assets/5bf01867-f1a3-4c78-9948-e51b4bba05e5" />

<img width="1110" height="223" alt="image" src="https://github.com/user-attachments/assets/f30973e3-620c-4c72-9443-26edfd8246b8" />

<img width="1198" height="168" alt="image" src="https://github.com/user-attachments/assets/85cdd197-a76d-4c9c-accf-3947f5082880" />

<img width="1205" height="205" alt="image" src="https://github.com/user-attachments/assets/ffcd979f-99f7-43d7-afb0-4c5fb8a7d364" />

gonna use:
```
function tradebench {
   $cmd = "python trade.py $args"
   Measure-Command { Invoke-Expression $cmd | Out-Default } | Select -Expand TotalMilliseconds
}
```

<img width="724" height="368" alt="image" src="https://github.com/user-attachments/assets/6a749368-9499-4711-809b-bd394fbaeb11" />

<img width="819" height="93" alt="image" src="https://github.com/user-attachments/assets/d6f12eac-b815-4c18-a859-e5541827b712" />

<img width="755" height="91" alt="image" src="https://github.com/user-attachments/assets/f9c26c24-ead9-4c9d-ad76-ef52351e7419" />

<img width="1206" height="660" alt="image" src="https://github.com/user-attachments/assets/0a18aeb2-7df2-48b4-8a19-fb0f83db3b4d" />

<img width="806" height="701" alt="image" src="https://github.com/user-attachments/assets/e64bed5c-9699-42e7-90aa-577e5526d54f" />

<img width="793" height="704" alt="image" src="https://github.com/user-attachments/assets/2cdb6919-1a92-48f5-bd7e-068dcb0acc80" />

# hacked tradegame to not load navroute
<img width="1176" height="168" alt="image" src="https://github.com/user-attachments/assets/d964f47a-9ea8-483f-9ff0-2f27ea0b6d89" />

# no navroute set in-game:
<img width="776" height="102" alt="image" src="https://github.com/user-attachments/assets/12066783-e8ff-4557-b56b-8557fe4e524d" />

# set target in-game, note there's a slight extra loading hitch as it loads the new journal
<img width="740" height="82" alt="image" src="https://github.com/user-attachments/assets/c0eb78d7-c1d6-4373-b345-8dbad55aa26f" />

kfsone commented:
And an import/update:
```
(.venv) [01/10/2026 14:17:13 |feat/trade-game| G:\Elite\Trade-Dangerous]
Win|pwsh> tradebench import -P journal -vv -ww  # no update yet
#0: Command line was: ['trade.py', 'import', '-P', 'journal', '-vv', '-ww']
#0: data_dir = G:\Elite\Trade-Dangerous\data
#0: db_path = G:\Elite\Trade-Dangerous\data\TradeDangerous.db
#0: db_config = G:\Elite\Trade-Dangerous\data\db_config.ini
#0: journal_path: windows: None
#0: elite journal path: C:\Users\oliver.smith\Saved Games\Frontier Developments\Elite Dangerous
#1: elite game journal load: <JournalLoad.DO_NOT_LOAD: 0>
#1: elite game json files: [<JsonFiles.MARKET: 'Market.json'>]
#1: Refreshing game state
#1: JournalLoad.DO_NOT_LOAD; cleared journals
#0: _load_journals took 0.000711 seconds
#1: loading json file: C:\Users\oliver.smith\Saved Games\Frontier Developments\Elite Dangerous\Market.json
#0: _load_json_files took 0.002238 seconds
#0: EliteGame refresh took 0.004781 seconds
NOTE: Alrai/Surveyor’s Reach (#129032283) 2026-01-10 09:28:37+00:00: (352) items
NOTE: Data already imported.
891.5498
(.venv) [01/10/2026 14:18:32 |feat/trade-game| G:\Elite\Trade-Dangerous]
Win|pwsh> tradebench import -P journal -vv -ww  # opened market in-game
#0: Command line was: ['trade.py', 'import', '-P', 'journal', '-vv', '-ww']
#0: data_dir = G:\Elite\Trade-Dangerous\data
#0: db_path = G:\Elite\Trade-Dangerous\data\TradeDangerous.db
#0: db_config = G:\Elite\Trade-Dangerous\data\db_config.ini
#0: journal_path: windows: None
#0: elite journal path: C:\Users\oliver.smith\Saved Games\Frontier Developments\Elite Dangerous
#1: elite game journal load: <JournalLoad.DO_NOT_LOAD: 0>
#1: elite game json files: [<JsonFiles.MARKET: 'Market.json'>]
#1: Refreshing game state
#1: JournalLoad.DO_NOT_LOAD; cleared journals
#0: _load_journals took 0.000856 seconds
#1: loading json file: C:\Users\oliver.smith\Saved Games\Frontier Developments\Elite Dangerous\Market.json
#0: _load_json_files took 0.001633 seconds
#0: EliteGame refresh took 0.004177 seconds
NOTE: Alrai/Surveyor’s Reach (#129032283) 2026-01-10 22:18:43+00:00: (352) items
NOTE: Updating: previous entry: 2026-01-10 09:28:37+00:00/12 hrs
#0: item: {'id': 128049152, 'Name': '$platinum_name;', 'Name_Localised': 'Platinum', 'Category': '$MARKET_category_metals;', 'Category_Localised': 'Metals', 'BuyPrice': 0, 'SellPrice': 57273, 'MeanPrice': 55505,
'StockBracket': 0, 'DemandBracket': 3, 'Stock': 0, 'Demand': 15209, 'Consumer': True, 'Producer': False, 'Rare': False}
#0: item: {'id': 128049153, 'Name': '$palladium_name;', 'Name_Localised': 'Palladium', 'Category': '$MARKET_category_metals;', 'Category_Localised': 'Metals', 'BuyPrice': 0, 'SellPrice': 53896, 'MeanPrice':
50356, 'StockBracket': 0, 'DemandBracket': 3, 'Stock': 0, 'Demand': 20467, 'Consumer': True, 'Producer': False, 'Rare': False}
#0: item: {'id': 128049154, 'Name': '$gold_name;', 'Name_Localised': 'Gold', 'Category': '$MARKET_category_metals;', 'Category_Localised': 'Metals', 'BuyPrice': 0, 'SellPrice': 51032, 'MeanPrice': 47113,
'StockBracket': 0, 'DemandBracket': 3, 'Stock': 0, 'Demand': 24784, 'Consumer': True, 'Producer': False, 'Rare': False}
#0: item: {'id': 128049155, 'Name': '$silver_name;', 'Name_Localised': 'Silver', 'Category': '$MARKET_category_metals;', 'Category_Localised': 'Metals', 'BuyPrice': 0, 'SellPrice': 40917, 'MeanPrice': 36553,
'StockBracket': 0, 'DemandBracket': 3, 'Stock': 0, 'Demand': 222576, 'Consumer': True, 'Producer': False, 'Rare': False}
#0: item: {'id': 128049156, 'Name': '$bertrandite_name;', 'Name_Localised': 'Bertrandite', 'Category': '$MARKET_category_minerals;', 'Category_Localised': 'Minerals', 'BuyPrice': 13094, 'SellPrice': 13053,
'MeanPrice': 18210, 'StockBracket': 0, 'DemandBracket': 0, 'Stock': 0, 'Demand': 0, 'Consumer': False, 'Producer': False, 'Rare': False}
#0: item: {'id': 128049157, 'Name': '$indite_name;', 'Name_Localised': 'Indite', 'Category': '$MARKET_category_minerals;', 'Category_Localised': 'Minerals', 'BuyPrice': 7748, 'SellPrice': 7723, 'MeanPrice':
11044, 'StockBracket': 0, 'DemandBracket': 0, 'Stock': 0, 'Demand': 0, 'Consumer': False, 'Producer': False, 'Rare': False}
...
```

Tromador commented:
This is looking much better, just one more note.

### Possible Logic Inversion `get_status()`

In EliteGame.get_status():

- cargo_load is computed from the Cargo journal event (sum of Count).
- cargo_space (capacity) is taken from the Loadout journal event (CargoCapacity).

But in require_game_data() the event requirements are currently flipped:
```
if cargo or cargo_cap:
    events.add(("Cargo",))

if cargo or cargo_use:
    events.add(("Loadout",))
```
Or at least that's how it looks to me. They need swapping to.
```
if cargo or cargo_use:
    events.add(("Cargo",))

if cargo or cargo_cap:
    events.add(("Loadout",))
```
And it might be an idea to avoid surprised by making it tolerant of `None`.
```
cargo_space = max(status.cargo_space or 0, 0)
cargo_load  = max(status.cargo_load  or 0, 0)
```

kfsone commented:
> This is looking much better, just one more note.
> ```
> if cargo or cargo_cap:
>     events.add(("Cargo",))
> 
> if cargo or cargo_use:
>     events.add(("Loadout",))
> ```
> 
> Or at least that's how it looks to me. They need swapping to.

Exactly right, fixed.

> And it might be an idea to avoid surprised by making it tolerant of `None`.
> 
> ```
> cargo_space = max(status.cargo_space or 0, 0)
> cargo_load  = max(status.cargo_load  or 0, 0)
> ```

Hmm - I'm wary of that, since it lets a bug self-squelch. We've added a call to `require_game_data` in the trust that it will do our None checks for us, and I already stopped short of:

<img width="1036" height="152" alt="image" src="https://github.com/user-attachments/assets/db3b4137-0a62-4c44-8dd2-9bb192bf5ed5" />

(I changed the parameter names for consistency, apparently when I was writing the method I got distracted by aligning the ":"s, "="s, and "#"s *blush*)

see: https://github.com/eyeonus/Trade-Dangerous/pull/252/commits/d41739508ea8bdf43a0d55fbb67a8acc1b8636ab
for the changes pre-squash

Tromador commented:
### Logic Inversion: Fixed ✅

Here's my problem with asserts, pretty much ever, but especially in a situation like this.

`python -O` strips them out. Even if people don't use `-O` they might have `PYTHONOPTIMIZE=1` set in their env. That immediately drops ALL assertions and will make your `None` checks disappear.

So yes, don't self-squelch bugs, I hear you, but asserts aren't the way to deal with it, instead perhaps make the guard non-optional:
- Replace assert with an explicit runtime check that raises the same friendly `GameDataException` guidance you already have (or maybe a `CommandLineError` in `trade_cmd`).
Example shape:
```
if status.cargo_space is None:
    raise GameDataException("... guidance ...")
```

Or better: push it into `require_game_data(... status_fields=...)` so all commands benefit.

I mean, I have no idea really how likely it is folk are using optimise flags, but I think it's better to assume the worst.

kfsone committed:
review feedback: additional None checks

kfsone commented:
> Or better: push it into `require_game_data(... status_fields=...)` so all commands benefit.

Ahh! I see it; I think this turns out nicely. Couple self-checks:

<img width="700" height="379" alt="image" src="https://github.com/user-attachments/assets/f774120b-8033-47b5-8c34-1bee3a53c13a" />
(intentional typo is rejected)

Tromador committed:
Replace assertion with RuntimeError for max_journals

If it needs to be a real guard, it needs to not be an assertion.

Tromador committed:
Add GameDataError for data inconsistency handling

Introduced GameDataError exception to handle inconsistencies in imported or journal data.

Tromador committed:
Raise error for inconsistent cargo load

Add error handling for negative cargo space after deduction.

Tromador committed:
Fix class definition for GameDataError

Tromador committed:
Add GameDataError exception to trade_cmd.py

Tromador commented:
Given you said we could mess with your PRs. I did (see me learning stuff). I wouldn't want to do it for big fixes though, these were tiny.

I swapped an assert (again... -O etc blahblah) for a proper guard and added a check for a negative cargo space, which should never ever happen, unless the journal has gone bad or there's a change which isn't quite up to date... or something caused by wrong numbers in the journal - together with a new exception to handle such events.

Tromador commented:
Please could you post something to the [forum page](https://forums.frontier.co.uk/threads/trade-dangerous-est-2015-power-users-highly-configurable-trade-optimizer.441509/) explaining the new functionality.

We really need to update our documentation, it's badly out of date, but at least if it's on the forum!
